debian/patches/0080-add-tonemapx-filter.patch

```diff
@@ -0,0 +1,5601 @@
++            g1x4b = _mm_add_epi32(g1x4b, _mm_mullo_epi32(vx4b, _mm_set1_epi32(cgv)));
++            g1x4b = _mm_add_epi32(g1x4b, rndx4);
++            g1x4b = _mm_srai_epi32(g1x4b, in_sh);
++            g1x4b = av_clip_int16_sse(g1x4b);
++
++            b0x4b = _mm_add_epi32(b0x4b, _mm_mullo_epi32(ux4b, _mm_set1_epi32(cbu)));
++            b0x4b = _mm_add_epi32(b0x4b, rndx4);
++            b0x4b = _mm_srai_epi32(b0x4b, in_sh);
++            b0x4b = av_clip_int16_sse(b0x4b);
++
++            b1x4b = _mm_add_epi32(b1x4b, _mm_mullo_epi32(ux4b, _mm_set1_epi32(cbu)));
++            b1x4b = _mm_add_epi32(b1x4b, rndx4);
++            b1x4b = _mm_srai_epi32(b1x4b, in_sh);
++            b1x4b = av_clip_int16_sse(b1x4b);
++
++            tonemap_int32x4_sse(r0x4a, g0x4a, b0x4a, r, g, b,
++                                params->lin_lut, params->tonemap_lut, params->delin_lut,
++                                params->coeffs, params->ocoeffs, params->desat, params->rgb2rgb_coeffs,
++                                params->rgb2rgb_passthrough);
++            tonemap_int32x4_sse(r1x4a, g1x4a, b1x4a, r1, g1, b1,
++                                params->lin_lut, params->tonemap_lut, params->delin_lut,
++                                params->coeffs, params->ocoeffs, params->desat, params->rgb2rgb_coeffs,
++                                params->rgb2rgb_passthrough);
++            tonemap_int32x4_sse(r0x4b, g0x4b, b0x4b, &r[4], &g[4], &b[4],
++                                params->lin_lut, params->tonemap_lut, params->delin_lut,
++                                params->coeffs, params->ocoeffs, params->desat, params->rgb2rgb_coeffs,
++                                params->rgb2rgb_passthrough);
++            tonemap_int32x4_sse(r1x4b, g1x4b, b1x4b, &r1[4], &g1[4], &b1[4],
++                                params->lin_lut, params->tonemap_lut, params->delin_lut,
++                                params->coeffs, params->ocoeffs, params->desat, params->rgb2rgb_coeffs,
++                                params->rgb2rgb_passthrough);
++
++            r0ox8 = _mm_lddqu_si128((const __m128i_u *)r);
++            g0ox8 = _mm_lddqu_si128((const __m128i_u *)g);
++            b0ox8 = _mm_lddqu_si128((const __m128i_u *)b);
++
++            roax4 = _mm_cvtepi16_epi32(r0ox8);
++            goax4 = _mm_cvtepi16_epi32(g0ox8);
++            boax4 = _mm_cvtepi16_epi32(b0ox8);
++
++            robx4 = _mm_unpackhi_epi16(r0ox8, zero128);
++            gobx4 = _mm_unpackhi_epi16(g0ox8, zero128);
++            bobx4 = _mm_unpackhi_epi16(b0ox8, zero128);
++
++            yoax4 = _mm_mullo_epi32(roax4, _mm_set1_epi32(cry));
++            yoax4 = _mm_add_epi32(yoax4, _mm_mullo_epi32(goax4, _mm_set1_epi32(cgy)));
++            yoax4 = _mm_add_epi32(yoax4, _mm_mullo_epi32(boax4, _mm_set1_epi32(cby)));
++            yoax4 = _mm_add_epi32(yoax4, _mm_set1_epi32(out_rnd));
++            yoax4 = _mm_srai_epi32(yoax4, 21);
```

Comment:
Use `out_sh` or hardcoded `21`?

Reply:
For arm neon, the right shift bit has to be a compile-time constant so it is hardcoded to 21 for all 8bit outputs. SSE and AVX does not have this constraint but I still prefer to make it a compile-time constant as the output depth is also a constant.

Reply:
If we want to reduce magic numbers, we can define it as macro in vf_tonemapx.h

Reply:
I saw those. Better add a comment near the hardcoded shift val to show where it comes from.

```
//    const int in_sh2 = 16 - in_depth; //<= Remove
...
            if (in_depth == 10) {
                // shift to low10bits for 10bit input
                // shift bit has to be compile-time constant
                y0x8 = vshrq_n_u16(y0x8, 6); // (16 - 10)
                y1x8 = vshrq_n_u16(y1x8, 6);
                uvx8 = vshrq_n_u16(uvx8, 6);
            }
```

Using macros is also acceptable, but it's a bit overkill.